TSDB support for Lens, TSVB and Timelion

flash1293 opened the pull request:
Fixes https://github.com/elastic/kibana/issues/131517
Fixes https://github.com/elastic/kibana/issues/131518
Fixes https://github.com/elastic/kibana/issues/136403
Fixes https://github.com/elastic/kibana/issues/136404

This PR is implementing the first parts of supporting TSDB in Lens, TSVB and Timelion.

## How to test

* Check out https://github.com/elastic/elasticsearch/tree/kibana-downsampling-branch next to your kibana working copy
* Start ES with yarn es source instead of snapshot (this will build the given branch)
* Clone https://github.com/csoulios/downsampling-demo
* Run the commands in [downsample-index.es](http://downsample-index.es/) (indexing the actual data is too much for dev console, so you have to do this via curl curl -XPOST http://elastic:changeme@localhost:9200/sample-01/_bulk\?pretty -H 'Content-Type: application/json' --data-binary @sample-01-raw.json)
* After running all commands, there should be sample-01-rollup which has the rolled up data
* Create a data view `sample-01` for testing TSDB without rollups and a data view `sample-01,sample-01-rollup` for testing TSDB with rollups

## Lens

* Implement new field type picker
* Add counter and gauge field types to the list as separate types
* When dragging a field that is also defined on rollup indices, do not default to median, instead default to average
* When rendering the function list in the dimension editor, split the function list into "safe" functions that work with rolled up data too and "unsafe" functions that won't work
* When a field is defined both 

<img width="762" alt="Screenshot 2022-08-17 at 16 56 42" src="https://user-images.githubusercontent.com/1508364/185172809-e47f73e0-c58d-4aae-a220-41f3e900327a.png">
<img width="449" alt="Screenshot 2022-08-17 at 16 59 17" src="https://user-images.githubusercontent.com/1508364/185173276-089138aa-5abe-4d4f-8883-8fabdc7fc8a9.png">
<img width="542" alt="Screenshot 2022-08-17 at 16 59 22" src="https://user-images.githubusercontent.com/1508364/185173286-ef8b2f97-c82f-441a-a4ed-5a5d31a029ed.png">

## TSVB

If the current data view is hitting rolled up data, date histogram aggregations are always using fixed_interval instead of calendar_interval and UTC instead of the configured timezone.

This can be checked via the inspector.

It should not influence TSVB charts which are executed on index patterns not hitting rolled up timeseries indices.

## Timelion

Like TSVB, if the current data view is hitting rolled up data, date histogram aggregations are always using fixed_interval instead of calendar_interval and UTC instead of the configured timezone.

This can't be checked via inspector as Timelion isn't reporting executed requests - the easiest way to test is to add a `console.log` for the sent search in https://github.com/elastic/kibana/blob/1681ebe95e61d0e4cadeaa49ce18605b8474ca22/src/plugins/vis_types/timelion/server/series_functions/es/index.js#L131


## Split out intentionally

* Functional tests (this is not merged yet on Elasticsearch side, gonna follow up in another PR)
* Shard failure handling
* Splitting function list in the dropdown of moving average and differences
* Some kind of indication of safe and unsafe functions in Lens formula
* UTC and fixed interval enforcement in Lens and agg based visualizations: https://github.com/elastic/kibana/issues/139081

elasticmachine commented:
Pinging @elastic/kibana-vis-editors @elastic/kibana-vis-editors-external (Team:VisEditors)

KOTungseth commented:
<img width="449" alt="185173276-089138aa-5abe-4d4f-8883-8fabdc7fc8a9" src="https://user-images.githubusercontent.com/38990715/185419766-e5463488-1601-47ad-b43d-e55d8794ec38.png">

Since the results are partial instead of the application, would it make sense to rename `Partially applicable functions` to `Partial results functions`?

<img width="542" alt="185173286-ef8b2f97-c82f-441a-a4ed-5a5d31a029ed" src="https://user-images.githubusercontent.com/38990715/185420111-e6c14595-1a48-48f5-afc9-1891ed64be72.png">

How about this for the tooltip:

`The full time range of your data is unsupported by the Partial results functions` (if this is what we want to go with)`. For example, when you use rolled up historical data. You can still use these functions, but the results are partial and the visualization displays a warning.`

ppisljar reviewed (approved):
LGTM

MichaelMarcialis commented:
> <img alt="Screenshot 2022-08-17 at 16 59 17" width="449" src="https://user-images.githubusercontent.com/1508364/185173276-089138aa-5abe-4d4f-8883-8fabdc7fc8a9.png">
> 
> <img alt="Screenshot 2022-08-17 at 16 59 22" width="542" src="https://user-images.githubusercontent.com/1508364/185173286-ef8b2f97-c82f-441a-a4ed-5a5d31a029ed.png">

When was it decided that we would be splitting the quick functions list in this manner? I'm personally not a fan of this direction. I would much rather we take a similar approach to how we handle incompatible quick functions, where we keep a single unified function list but visually offset the incompatible functions (with the yellow dot) and provide a tooltip there. Thoughts?

flash1293 commented:
This is different than the “incompatible” dot, as it’s possible to use this function with the current field, but it might produce partial data. Requiring the user to hover over the dot to learn about this seems too subtle to me, but if you and @ghudgins  think it would be the way to go I’m happy with that too. What about a different kind of indicator for this?

andrewctate commented:
I get this error when I run `yarn es source`
<img width="936" alt="Screen Shot 2022-08-18 at 5 27 58 PM" src="https://user-images.githubusercontent.com/315764/185506188-82c20838-32bd-4293-8887-1e7367eb58bb.png">

These are the contents of my `.idea` folder
<img width="662" alt="Screen Shot 2022-08-18 at 5 34 25 PM" src="https://user-images.githubusercontent.com/315764/185506918-404333b8-eb36-45b2-b86c-01fc74a2181c.png">


I tried installing Eclipse and using it to open `Elasticsearch`. I also tried adding an empty XML file where it's looking for it, but it just keeps looking for more and more XML files.

Do you know how to solve this @flash1293 ?

flash1293 commented:
Unfortunately I don't, I didn't run into this error. @spalger do you know what's going on here? https://github.com/elastic/kibana/pull/139020#issuecomment-1220046143

spalger commented:
I have very limited experience with `yarn es source` and I think very few people use it. Are there modifications in the ES repo that might not be fully applied? I wouldn't be surprised if the "install_source" stuff should be reworked.

MichaelMarcialis commented:
> I wrote my comment before you added the last sentence - agreed with that 😁 no concerns with an indicator per function in a unified list but it should be different than the existing one.

Cool. EUI happens to have a `partial` icon. We could append that icon to the functions in question, similar to how we append the yellow "incompatible" dots. @ghudgins, did you want to weigh in on this direction versus splitting the quick function list?

andrewctate reviewed (approved):
I tested all three apps and went through the code. LGTM—just had a few questions.

### 1.
> "unsafe" functions that won't work

Why do we allow these to be selected at all?

### 2.
I verified that both TSVB and Timelion switch their requests to use UTC as soon as they're looking at rolled-up data. However, in my case, they both attached fixed_interval regardless of whether or not they were dealing with rolled up data. Could have been my visualization configuration, but thought it was worth mentioning since I didn't figure out how to make them use calendar interval for non-rolled-up data.

### 3.
Am I correct in understanding that we are ultimately [relying on the field caps API](https://github.com/elastic/kibana/pull/137677/files#diff-7fd8f81570bb34b0ee8771f29f9ace6925c5cfa7144f75ede8bdc09275160f19R145) to report a list of possible fixed intervals for a rolled-up field?

All my other comments will be addressed by your follow-up list 👍

ghudgins commented on x-pack/plugins/lens/public/indexpattern_datasource/dimension_panel/dimension_editor.tsx:
```diff
@@ -516,6 +520,21 @@ export function DimensionEditor(props: DimensionEditorProps) {
+  if (softRestrictedSideNavItems.length > 0) {
+    sideNavItems = sideNavItems.filter(
+      (navItem) =>
+        !referencedField ||
+        !referencedField.softRestrictions ||
+        !referencedField.softRestrictions[navItem.id!]
```
>Why do we allow these to be selected at all?

Because the time range can change and they can suddenly work. The time range dictates if the indexes that are returned are downsampled or not

flash1293 commented on x-pack/plugins/lens/public/indexpattern_datasource/datapanel.tsx:
```diff
@@ -311,9 +313,11 @@ export const InnerIndexPatternDataPanel = function InnerIndexPatternDataPanel({
-  );
+  const availableFieldTypes = uniq([
+    ...uniq(allFields.map(getFieldType)).filter((type) => type in fieldTypeNames),
+    // always include current selection - there might be no match for an existing type filter on data view switch
```
> This is so that the user doesn't lose the ability to remove the filter

Spot on, I will adjust the comment

flash1293 commented on x-pack/plugins/lens/public/indexpattern_datasource/operations/definitions/percentile.tsx:
```diff
@@ -89,7 +89,11 @@ export const percentileOperation: OperationDefinition<
+    if (
+      supportedFieldTypes.includes(fieldType) &&
+      aggregatable &&
+      (!aggregationRestrictions || aggregationRestrictions.percentiles)
```
It's a slightly weird API - if `aggregationRestrictions` exists, it means not all aggs can be applied. If `aggregationRestrictions.percentiles` is set, it means it can be applied.

flash1293 commented on x-pack/plugins/lens/public/indexpattern_datasource/dimension_panel/dimension_editor.tsx:
```diff
@@ -516,6 +520,21 @@ export function DimensionEditor(props: DimensionEditorProps) {
       };
     }
   );
+  const softRestrictedSideNavItems = sideNavItems.filter(
```
Very good point, gonna do that.

flash1293 commented:
Thanks for the review, addressed all comments.

> Why do we allow these to be selected at all?

As Graham noted, this is a "soft" restriction because in 99% of the cases there is also recent non-rolled up data for which it makes sense to apply these functions. So we allow users to pick it and live with the warnings. Note that in case of the sample data as described above the rolled-up and regular data is in the same time range but that's rather unusual. In most cases the rolled up data will be in another time range (e.g. last week won't be rolled up, but the month before that will), so the user won't actually get a warning as long as looking at the last week.

> However, in my case, they both attached fixed_interval regardless of whether or not they were dealing with rolled up data

As long as the selected data view includes at least one index that's rolled up that's the expected behavior - it's difficult but possible to detect this situation on a per-query basis and adjust the interval based on this information, but we decided to not go with this as it will be extra confusing when suddenly crossing the boundary (for example looking at the last 5 days will give you calendar intervals in the current time zone, changing the time range to 8 days will suddenly give you fixed intervals in UTC). Data views which don't include rolled up data at all should not be affected (not sure which case you ran into - the described behavior is what I'm seeing locally)

This is how the function list looks now:
<img width="417" alt="Screenshot 2022-08-29 at 12 11 06" src="https://user-images.githubusercontent.com/1508364/187182488-dd6d4c16-16e8-4e04-a2e4-d44c54fddc95.png">
<img width="640" alt="Screenshot 2022-08-29 at 12 11 10" src="https://user-images.githubusercontent.com/1508364/187182493-ba91f6a1-2cb0-466e-be60-60119d4abad8.png">

Note that the "incompatible" dot will never show up together with the partial icon - if a field is incompatible with a function the current field is not transferred over so we do not know whether it will be partially applicable or not (as this depends on the choice the user didn't make yet)

MichaelMarcialis reviewed (approved):
Thanks for making that change to the quick function list, @flash1293! That looks much better to my eyes. I'm leaving a few comments below for your review, but nothing earth-shattering. Assuming those get addressed, I'm approving now so I don't hold you up further.

- I noticed that the "Search field names" input above the field list has a double border (CSS `box-shadow` in reality) effect happening in this PR that doesn't seem to appear in the `main` branch. Any chance to fix it so this artificial thickening of the input border doesn't happen?

![image](https://user-images.githubusercontent.com/3884767/187268452-2b37288a-56a9-465b-b126-cd4907b03d53.png)

- Since you mentioned that the quick functions' incompatible `dot` icons can't appear simultaneously with the newly implemented `partial` icon, would you be willing to try coloring the `partial` icon in the same yellow color we use for the incompatible `dot` icons? Assuming it doesn't look awful, that might help to both visually offset it from the text color and be consistent with the notion that it's a warning.

flash1293 commented:
It looks good IMHO

<img width="390" alt="Screenshot 2022-08-30 at 12 17 23" src="https://user-images.githubusercontent.com/1508364/187412379-af8d6b60-d1a4-46dd-acfc-4efc533b48f6.png">

ghudgins commented:
dots look good. sorry for the late 👀
